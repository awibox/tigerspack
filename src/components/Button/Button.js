import React from 'react';
import PropTypes from 'prop-types';
import { colors } from '../../utils/colors';
import { defaultStyles } from '../../utils/defaultStyles';

const Button = (props) => {
  const {
    children,
    icon,
    indent,
    theme,
    type,
    ...otherProps
  } = props;
  const palette = colors[theme] ? colors[theme] : colors.primary;
  const size = props.size > 0 && props.size < 6 ? props.size : defaultStyles.buttonSize;
  // Dynamic styles
  const dynamicStyles = {};
  switch (type) {
    case 'text': {
      dynamicStyles.background = 'transparent';
      dynamicStyles.backgroundHover = palette.color;
      dynamicStyles.border = '1px solid transparent';
      dynamicStyles.color = palette.color;
      dynamicStyles.colorHover = palette.hoverText;
      break;
    }
    case 'outline': {
      dynamicStyles.background = 'transparent';
      dynamicStyles.border = `1px solid ${palette.color}`;
      dynamicStyles.backgroundHover = palette.color;
      dynamicStyles.color = palette.color;
      dynamicStyles.colorHover = palette.hoverText;
      break;
    }
    default: {
      dynamicStyles.background = palette.color;
      dynamicStyles.border = '1px solid transparent';
      dynamicStyles.backgroundHover = palette.hoverColor;
      dynamicStyles.color = palette.text;
      dynamicStyles.colorHover = palette.text;
      break;
    }
  }
  // Styles
  const styles = {
    button: {
      boxSizing: 'border-box',
      position: 'relative',
      outline: 'none',
      transition: defaultStyles.animation,
      display: 'flex',
      flexDirection: 'column',
      flexFlow: 'row',
      alignContent: 'center',
      alignItems: 'center',
      justifyContent: 'center',
      justifyItems: 'center',
      marginBottom: `${indent}px`,
      fontWeight: 500,
      backgroundImage: 'none',
      whiteSpace: 'nowrap',
      userSelect: 'none',
      cursor: 'pointer',
      lineHeight: 0,
      borderRadius: type !== 'rounded' ? defaultStyles.borderRadius : '20px',
      backgroundColor: dynamicStyles.background,
      border: dynamicStyles.border,
      color: dynamicStyles.color,
      fontSize: `${size * 2 + 9}px`,
      height: `${(size * 2 + 9) + (size * 5)}px`,
      padding: `${((size * 5) / 2) - 0.5}px ${size * 5}px ${((size * 5) / 2) + 0.5}px`,
      ':disabled': {
        opacity: 0.7,
        cursor: 'not-allowed',
        ':hover': {
          color: dynamicStyles.color,
          backgroundColor: dynamicStyles.background,
        },
      },
      ':hover': {
        color: dynamicStyles.colorHover,
        textDecoration: 'none',
        backgroundColor: dynamicStyles.backgroundHover,
      },
    },
    icon: {
      marginLeft: `-${(size + 5) / 2}px`,
      marginRight: children ? `${(size + 5)}px` : `-${(size + 5) / 2}px`,
    },
  };
  return (
    <button css={styles.button} {...otherProps}>
      {icon ? <span css={styles.icon}>{icon}</span> : ''}
      {children}
    </button>
  );
};

Button.propTypes = {
  theme: PropTypes.string,
  children: PropTypes.any,
  size: PropTypes.oneOf([1, 2, 3, 4, 5]),
  type: PropTypes.oneOf(['text', 'outline', 'rounded']),
  icon: PropTypes.any,
  indent: PropTypes.number,
};

Button.defaultProps = {
  indent: 0,
};

export default Button;
