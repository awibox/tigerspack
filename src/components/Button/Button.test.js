import React from 'react';
import { shallow } from 'enzyme';
import Button from '.';

const clickFn = jest.fn();

describe('Button', () => {
  it('should render correctly', () => {
    const component = shallow(<Button onClick={clickFn}>Test button</Button>);
    expect(component).toMatchSnapshot();
  });
  it('should call clickFn on button click', () => {
    const component = shallow(<Button onClick={clickFn}>Test button</Button>);
    component.simulate('click');
    expect(clickFn).toHaveBeenCalled();
  });
});
