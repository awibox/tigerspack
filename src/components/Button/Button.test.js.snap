// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render correctly 1`] = `
.button {
  border-radius: 4px !important;
  -moz-box-sizing: border-box !important;
  box-sizing: border-box !important;
  position: relative !important;
  font-size: 14px !important;
  outline: none !important;
  -webkit-transition: all 0.1s !important;
  -moz-transition: all 0.1s !important;
  transition: all 0.1s !important;
  display: inline-block !important;
  margin-bottom: 0px !important;
  font-weight: 500 !important;
  text-align: center !important;
  vertical-align: middle !important;
  cursor: pointer !important;
  background-image: none !important;
  border: 1px solid transparent !important;
  white-space: nowrap !important;
  padding: 6px 16px !important;
  line-height: 1.52857143 !important;
  -webkit-user-select: none !important;
  -moz-user-select: none !important;
  -ms-user-select: none !important;
  user-select: none !important;
  background-color: #3F88C5 !important;
  color: #fff !important;
}

.button:hover {
  text-decoration: none !important;
  background-color: #619DCF !important;
}

<div
  className="button"
  onClick={[MockFunction]}
>
  Test button
</div>
`;
